src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java

```diff
@@ -168,19 +153,28 @@ public void start() throws MuleException {
                            tokenResponse.getAccessToken(), tokenResponse.getRefreshToken(), tokenResponse.getExpiresIn());
             }
 
-            final DefaultResourceOwnerOAuthContext defaultUserState = (DefaultResourceOwnerOAuthContext) getContext();
             defaultUserState.setAccessToken(tokenResponse.getAccessToken());
             defaultUserState.setExpiresIn(tokenResponse.getExpiresIn());
             for (Entry<String, Object> customResponseParameterEntry : tokenResponse.getCustomResponseParameters().entrySet()) {
               defaultUserState.getTokenResponseParameters().put(customResponseParameterEntry.getKey(),
                                                                 customResponseParameterEntry.getValue());
             }
 
+            defaultUserState.setDancerState(HAS_TOKEN);
             updateResourceOwnerOAuthContext(defaultUserState);
             if (notifyListeners) {
               listeners.forEach(l -> l.onTokenRefreshed(defaultUserState));
             }
           });
+        })
+        .exceptionally(t -> {
+          defaultUserState.setDancerState(NO_TOKEN);
```

Comment:
not sure of the effort involved or if it's even possible, but it'd be great if this entire danceState thing could be handled in the base class instead of each dancer.